packages/graph/hash_graph/lib/interval-ops/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+    clippy::use_debug,
+    clippy::verbose_file_reads
+)]
+// Until we do optimization work, there is unlikely to be any reason to use unsafe code. When it
+// becomes necessary/desirable to allow for unsafe code, we should:
+// - enable miri checks in the CI for the relevant code
+// - swap this lint with `#![deny(unsafe_code)]` and only allow it in a few places unless, or until,
+//   it gets very verbose to do so.
+#![forbid(
+    unsafe_code,
+    reason = "Unsafe code has been disabled until a good argument has been put forward for its \
+              usage"
+)]
+#![allow(
+    clippy::module_name_repetitions,
+    reason = "This encourages importing `as` which breaks IDEs"
+)]
+
+mod bounds;
+#[cfg(feature = "canonical")]
+mod canonical;
+#[cfg(feature = "continuous")]
+mod continuous;
+mod interval;
+
+#[cfg(feature = "canonical")]
+pub use self::canonical::CanonicalInterval;
+#[cfg(feature = "continuous")]
+pub use self::continuous::ContinuousInterval;
+pub use self::{
+    bounds::{LowerBound, UpperBound},
+    interval::Interval,
+};
+
+#[inline(never)]
+fn invalid_bounds() -> ! {
+    panic!("interval lower bound must be less than or equal to its upper bound")
+}
+
+#[cfg(all(test, feature = "canonicalize"))]
+mod tests {
+    use core::ops::Bound;
+
+    use crate::{CanonicalInterval, ContinuousInterval, Interval};
+
+    #[test]
+    fn new() {
+        assert_eq!(
+            ContinuousInterval::from_range(1..=2).canonicalize(),
+            CanonicalInterval::from_range(1..3)
+        );
+        assert_eq!(
+            ContinuousInterval::from_range((Bound::Excluded(0), Bound::Excluded(3))).canonicalize(),
+            CanonicalInterval::from_range(1..3)
+        );
+        assert_eq!(
+            ContinuousInterval::from_range(i32::MIN..=i32::MAX).canonicalize(),
+            CanonicalInterval::from_range(i32::MIN..)
+        );
+        assert_eq!(
+            ContinuousInterval::from_range(..=i32::MAX).canonicalize(),
+            CanonicalInterval::from_range(..)
+        );
+        assert!(ContinuousInterval::from_range(1..1).is_empty());
+        assert!(
+            !ContinuousInterval::from_range((Bound::Excluded(i32::MAX), Bound::Unbounded))
+                .is_empty()
+        );
+        assert!(
+            ContinuousInterval::from_range((Bound::Excluded(i32::MAX), Bound::Unbounded))
+                .canonicalize()
+                .is_empty()
+        );
+        assert!(
+            ContinuousInterval::from_range((Bound::Excluded(i32::MAX), Bound::Excluded(i32::MAX)))
+                .is_empty()
+        );
+        assert!(
+            ContinuousInterval::from_range((Bound::Excluded(i32::MAX), Bound::Excluded(i32::MAX)))
+                .canonicalize()
+                .is_empty()
+        );
+    }
+
+    #[test]
+    #[rustfmt::skip]
+    fn test_merge() {
+        assert_eq!(ContinuousInterval::from_range(1..3) + ContinuousInterval::from_range(0..5), ContinuousInterval::from_range(0..5));
```

Comment:
Since we don't have a dedicated `union` operator, I want to point out that `+` could be interpreted as Minkowski addition (pairwise addition before union), https://math.stackexchange.com/a/659134. Could be slightly problematic, but likely not to cause confusion as-is

Reply:
The Minkowski Sum is defined on sets, but not on intervals. The main difference is that we don't store any values between the lower and upper bounds but only the bounds themselves. But generally, I agree that we should remove the operator overloading as using the functions are both, more expressive and more powerful.